### PR TITLE
[REVIEW] Initial version of non-groupby _take_last

### LINKED ISSUE #4732: [DISCUSSION] Remove pandas groupby.last() dependency from _take_last method
Currently, cumulative aggregation operations are tied to non-hash based groupbys and pandas through the process of grabbing the last row and explicit pandas mentions in several places, particularly in [_take_last](https://github.com/dask/dask/blob/master/dask/dataframe/core.py#L4179).

Handling the explicit pandas references is already a TODO, but I'd like to raise for discussion removing the usage of  `groupby.last()`. Using pd.DataFrame.groupby.last() with a dummy group to capture the last row of each column works because pandas groupby preserves the order of rows within each group. Hash-based cuDF groupbys do not make such a guarantees, making this approach incompatible with cuDF.

There are several potential ways to change this if there is interest, and they have varying performance penalties / cleanliness. All of the approaches proposed below have a performance penalty that increases with the number of columns.

The clearest ways are to either apply `pd.Series.last_valid_index` and then perform the selection via the underlying numpy array, or to essentially mimic the functionality of `last_valid_index` either explicitly or in a large lambda function expression.

Given the following example data, here are some loose time benchmarks of approaches to getting the same data:
```
import pandas as pd
nrows = 1000000 # one million rows
pdf = pd.DataFrame({'a':['abc']*(nrows),
                     'b':range(nrows),
                     'c':range(nrows),
                     'd':range(nrows)
                    })
```

Using the current approach appears to be the fastest, but causes the issues listed above:

```
%%time
group_dummy = np.ones(len(pdf.index))
print(pdf.groupby(group_dummy).last())
       a       b       c       d
1.0  abc  999999  999999  999999
CPU times: user 80 ms, sys: 4 ms, total: 84 ms
Wall time: 85.7 ms
```

Explicitly mimicking pandas's `groupby.last` method (with some slight changes to remove explicitly using the underlying numpy array) found [here](https://github.com/pandas-dev/pandas/blob/v0.24.2/pandas/core/groupby/groupby.py#L1288)
```
def last(x):
    x = x[x.notna()]
    if len(x) == 0:
        return np.nan
    return x.iloc[-1]

pd.DataFrame(pdf.apply(last)).T
     a       b       c       d
0  abc  999999  999999  999999
CPU times: user 312 ms, sys: 40 ms, total: 352 ms
Wall time: 352 ms
```

Using a large lambda expression that functionally does the same as using `last`:

```
%time print(pd.DataFrame(pdf.apply(lambda x: x[x.notna()].iloc[-1] if not x.isnull().all() else None)).T)
     a       b       c       d
0  abc  999999  999999  999999
CPU times: user 452 ms, sys: 32 ms, total: 484 ms
Wall time: 484 ms
```

Using `last_valid_index`:

```
%%time
ix = pdf.apply(pd.Series.last_valid_index)
print(pd.DataFrame(pdf.values[ix, np.arange(pdf.shape[1])],index=df.columns).T)
     a       b       c       d
0  abc  999999  999999  999999
CPU times: user 372 ms, sys: 88 ms, total: 460 ms
Wall time: 456 ms
```
I wasn't able to easily figure out a good way to avoid using numpy with `last_valid_index`, but it's likely feasible. If we can't do that, this approach does not resolve the issues in a way that GPU libraries like cuDF can leverage.

In general, all of these non-`groupby.last() ` based approaches have a performance penalty of about 4-5x at 1 million records (example above) of the dataset above, 10 million records (1 second vs 4-5 seconds), and 50 million records (5 seconds vs 20-25 seconds).

**The problem though, is that the performance penalty increases with the number of columns unfortunately**. So if we change the example above to have seven columns, at 10 million records the groupby approach is still ~1 second but the non-groupby approach takes ~7 seconds.

I'd like to hear people's thoughts on this topic, as we'd like to be able to support dask-based cumulative aggregations for GPU dataframes. Do people have a sense of whether this kind of performance tradeoff is reasonable?

### LINKED ISSUE #4743: [BUG] cumulative aggregations fail when there are empty partitions
### Description
As @mrocklin noticed, the cumulative aggregation functions will error when there are empty partitions, due to the `_take_last` function hitting an index error trying to grab the 0th element of an empty array.

### Example

```python
import pandas as pd
import dask.dataframe as dd
df = pd.DataFrame({'x': [1, 2, 3, 4, 5, 6, 7, 8]})
ddf = dd.from_pandas(df, npartitions=4)
ddf[ddf.x < 5].cumsum().compute()
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-5-f2ddb62cae63> in <module>()
----> 1 ddf[ddf.x < 5].cumsum().compute()

/Users/nickbecker/NVIDIA/dask/dask/base.pyc in compute(self, **kwargs)
    154         dask.base.compute
    155         """
--> 156         (result,) = compute(self, traverse=False, **kwargs)
    157         return result
    158 

/Users/nickbecker/NVIDIA/dask/dask/base.pyc in compute(*args, **kwargs)
    396     keys = [x.__dask_keys__() for x in collections]
    397     postcomputes = [x.__dask_postcompute__() for x in collections]
--> 398     results = schedule(dsk, keys, **kwargs)
    399     return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
    400 

/Users/nickbecker/NVIDIA/dask/dask/threaded.pyc in get(dsk, result, cache, num_workers, pool, **kwargs)
     74     results = get_async(pool.apply_async, len(pool._pool), dsk, result,
     75                         cache=cache, get_id=_thread_get_id,
---> 76                         pack_exception=pack_exception, **kwargs)
     77 
     78     # Cleanup pools associated to dead threads

/Users/nickbecker/NVIDIA/dask/dask/local.pyc in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    460                         _execute_task(task, data)  # Re-execute locally
    461                     else:
--> 462                         raise_exception(exc, tb)
    463                 res, worker_id = loads(res_info)
    464                 state['cache'][key] = res

/Users/nickbecker/NVIDIA/dask/dask/local.pyc in execute_task(key, task_info, dumps, loads, get_id, pack_exception)
    228     try:
    229         task, data = loads(task_info)
--> 230         result = _execute_task(task, data)
    231         id = get_id()
    232         result = dumps((result, id))

/Users/nickbecker/NVIDIA/dask/dask/core.pyc in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/Users/nickbecker/NVIDIA/dask/dask/optimization.pyc in __call__(self, *args)
    940                              % (len(self.inkeys), len(args)))
    941         return core.get(self.dsk, self.outkey,
--> 942                         dict(zip(self.inkeys, args)))
    943 
    944     def __reduce__(self):

/Users/nickbecker/NVIDIA/dask/dask/core.pyc in get(dsk, out, cache)
    147     for key in toposort(dsk):
    148         task = dsk[key]
--> 149         result = _execute_task(task, cache)
    150         cache[key] = result
    151     result = _execute_task(out, cache)

/Users/nickbecker/NVIDIA/dask/dask/core.pyc in _execute_task(arg, cache, dsk)
    117         func, args = arg[0], arg[1:]
    118         args2 = [_execute_task(a, cache) for a in args]
--> 119         return func(*args2)
    120     elif not ishashable(arg):
    121         return arg

/Users/nickbecker/NVIDIA/dask/dask/dataframe/core.py in apply_and_enforce(*args, **kwargs)
   3869     func = kwargs.pop('_func')
   3870     meta = kwargs.pop('_meta')
-> 3871     df = func(*args, **kwargs)
   3872     if is_dataframe_like(df) or is_series_like(df) or is_index_like(df):
   3873         if not len(df):

/Users/nickbecker/NVIDIA/dask/dask/dataframe/core.py in _take_last(a, skipna)
   4213         last_row = a.groupby(group_dummy).last()
   4214         if isinstance(a, pd.DataFrame):  # TODO: handle explicit pandas reference
-> 4215             return pd.Series(last_row.values[0], index=a.columns)
   4216         else:
   4217             return last_row.values[0]

IndexError: index 0 is out of bounds for axis 0 with size 0
```

### PR DESCRIPTION
**Pull request by beckernick**
**Summary of Changes**
- Refactors `_take_last` to remove the `groupby.last()` usage
    - Uses a two-tiered approach to find the last valid value in order to maintain speed. New approach is significantly faster than the groupby.last() approach for large data sizes and comparable at smaller sizes.
- Replaces explicit pandas references with a check for whether the object is dataframe-like and uses the appropriate Series constructor based on the types.

- [x] Existing tests passed
- [ ] Added and passed test for empty partitions scenario
- [x] Passes `flake8 dask`

This closes #4732 and will also close #4743.

### CODE REVIEW

File: dask/dataframe/core.py [hunk 1]
```diff
@@ -4204,17 +4204,37 @@ def _take_last(a, skipna=True):
         Whether to exclude NaN
 
     """
+
+    def _collect_series_last_valid(s):
+        size = len(s)
+        for i in range(size - 1, max(size - 11, 0) - 1, -1):
```
Comment (mrocklin) [inline dask/dataframe/core.py]: I recommend `range(0, min(10, len(s)))` instead and use `-i` below.  I suspect that this will be easier for future reviewers to follow than the negative logic here.

File: dask/dataframe/core.py [hunk 2]
```diff
@@ -4204,17 +4204,37 @@ def _take_last(a, skipna=True):
...
+            if not nonnull.empty:
+                return nonnull.iloc[-1]
+        return None
+
+    def _collect_frame_last_valid(frame):
+        last_valid_rows = {}
+        for col in a.columns:
+            last_val = _collect_series_last_valid(a[col])
+            last_valid_rows[col] = last_val
+        return last_valid_rows
```
Comment (mrocklin) [inline dask/dataframe/core.py]: To reduce indirection, I recommend that we replace this function definition with a dictionary comprehension lower down.

File: dask/dataframe/core.py [hunk 3]
```diff
@@ -4204,17 +4204,37 @@ def _take_last(a, skipna=True):
...
-        if isinstance(a, pd.DataFrame):  # TODO: handle explicit pandas reference
-            return pd.Series(last_row.values[0], index=a.columns)
+        # in each column
+        if is_dataframe_like(a):
+            last_valid_rows = _collect_frame_last_valid(a)
+            return pd.Series(last_valid_rows, index=a.columns)
         else:
-            return last_row.values[0]
+            last_valid = _collect_series_last_valid(a)
+            return last_valid
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Maybe just 

```python
return _last_valid(a)
```

File: dask/dataframe/core.py [hunk 4]
```diff
@@ -4204,17 +4204,37 @@ def _take_last(a, skipna=True):
...
         # take last valid value excluding NaN, NaN location may be different
-        # in each columns
-        group_dummy = np.ones(len(a.index))
-        last_row = a.groupby(group_dummy).last()
-        if isinstance(a, pd.DataFrame):  # TODO: handle explicit pandas reference
-            return pd.Series(last_row.values[0], index=a.columns)
+        # in each column
+        if is_dataframe_like(a):
+            last_valid_rows = _collect_frame_last_valid(a)
+            return pd.Series(last_valid_rows, index=a.columns)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Maybe instead 

```python
return pd.Series({c: _last_valid(a[c]) for c in a.columns}, columns=a.columns)
```

This would remove the need for the function definition above.

File: dask/dataframe/core.py [hunk 5]
```diff
@@ -4204,17 +4204,37 @@ def _take_last(a, skipna=True):
         Whether to exclude NaN
 
     """
+
+    def _collect_series_last_valid(s):
+        size = len(s)
```
Comment (mrocklin) [inline dask/dataframe/core.py]: If this is only going to be used once then lets just inline it.

File: dask/dataframe/core.py (hunk 1 above)
Comment (beckernick) [reply dask/dataframe/core.py]: Agreed.

File: dask/dataframe/core.py (hunk 5 above)
Comment (beckernick) [reply dask/dataframe/core.py]: Used twice, so will keep as a function consistent with other suggestions in your review.

File: dask/dataframe/core.py (hunk 1 above)
Comment (beckernick) [reply dask/dataframe/core.py]: Starting range with `1` as otherwise we would grab the first row in the column unintentionally.

File: dask/dataframe/core.py (hunk 5 above)
Comment (mrocklin) [reply dask/dataframe/core.py]: I meant the definition of `size = len(s)`

```suggestion
```

File: dask/dataframe/core.py [hunk 6]
```diff
@@ -4204,17 +4204,28 @@ def _take_last(a, skipna=True):
         Whether to exclude NaN
 
     """
+
+    def _last_valid(s):
+        size = len(s)
+        for i in range(1, min(10, size)):
```
Comment (mrocklin) [inline dask/dataframe/core.py]:
```suggestion
        for i in range(1, min(10, len(s))):
```

**Comment by beckernick** [comment]
I'd still like to remove the explicit `pd.Series` if possible as I think it may cause cumulative aggregations to break for cudf eventually when we create DataFrame methods. For now, I believe it shouldn't break anything in cudf as cumulative aggregations are only at the Series level.

Is there a general way to do this in dask or do you think the best approach would be to look at `_cum_agg` (the only place this function is used) to see if it really needs this information as a Series?

**Comment by mrocklin** [comment]
> I'd still like to remove the explicit pd.Series if possible as I think it may cause cumulative aggregations to break for cudf eventually when we create DataFrame methods

Right, so it sounds like we either need a way to get a Series type from a DataFrame or Series.  This might be with something like 

```python
if is_series_like(obj):
    typ = type(obj)
elif is_dataframe_like(obj):
    typ = typ(obj.iloc[0])
```

Alternatively, we could ask cudf to support the addition of Pandas series objects, which probably isn't a bad idea anyway.

**Comment by beckernick** [comment]
Would the first approach you've listed make cudf a dependency of dask? I'd think this is not something we would want to do. I'm not sure if storing the type in a variable lets us do anything downstream to create the correct Series (cudf or pandas) without bringing in cudf, but I may be unaware of a function/method that would help in this (perhaps some kind of constructor wrapper based on the dataframe backend, maybe?). Would love to learn more.

**Comment by mrocklin** [comment]
> Would the first approach you've listed make cudf a dependency of dask? I'd think this is not something we would want to do. I'm not sure if storing the type in a variable lets us do anything downstream to create the correct Series (cudf or pandas) without bringing in cudf, but I may be unaware of a function/method that would help in this

No, it would not force a dependency of cudf.  This would grab the type of the object we were given dynamically at runtime.  The only way the type will be a `cudf.Series` object is if the input was a cudf object, so we'll obviously have already imported cudf at that time.

That being said, making cudf robust to interacting with pandas series objects is probably a good thing to do anyway.

**Comment by beckernick** [comment]
I agree. Making cudf robust to interacting with pandas does sound like a good idea.

To make sure I'm following what you mean, Are you saying that something like:

```python
if is_dataframe_like(obj):
    typ = typ(obj.iloc[0])
    if typ is cudf like:
        cudf.Series(obj)
```
Will be fine because the only situation in which the type would be cudf.Series is when we already have cudf and can thus use cudf.Series. Which is also why we couldn't use isinstance(obj, cudf.Series) here because we will only have cudf.Series in that scenario? If so, we're now on the same page!

**Comment by mrocklin** [comment]
I'm saying this

```python
if is_series_like(obj):
    typ = type(obj)
elif is_dataframe_like(obj):
    typ = type(obj.iloc[0])

return typ(out, index=a.columns)
```

**Comment by beckernick** [comment]
Wow. Learned something great here. I was not aware that type had this kind of behavior.

**Comment by mrocklin** [comment]
```python
s = pd.Series([1, 2 ,3])
typ = type(s)
assert typ is pd.Series
```

so this is the same:

```python 
pd.Series([1, 2, 3]) == typ([1, 2, 3])
```

File: dask/dataframe/core.py [hunk 7]
```diff
@@ -4204,17 +4204,28 @@ def _take_last(a, skipna=True):
...
         # take last valid value excluding NaN, NaN location may be different
-        # in each columns
-        group_dummy = np.ones(len(a.index))
-        last_row = a.groupby(group_dummy).last()
-        if isinstance(a, pd.DataFrame):  # TODO: handle explicit pandas reference
-            return pd.Series(last_row.values[0], index=a.columns)
+        # in each column
+        if is_dataframe_like(a):
+            # create Series of appropriate backend dataframe library
+            series_typ = type(a.iloc[0])
```
Comment (mrocklin) [inline dask/dataframe/core.py]: Hrm, this will fail if there are empty partitions, as in this example.

```python
import pandas as pd
import dask.dataframe as dd
df = pd.DataFrame({'x': [1, 2, 3, 4, 5, 6, 7, 8]})
ddf = dd.from_pandas(df, npartitions=4)
ddf[ddf.x < 5].cumsum().compute()
```

(sorry for not noticing this before)

Maybe `type(a.head(0).sum())` ?  Actually no, cudf doesn't support this.  We'll need to find some cheap way to get the Series type of a DataFrame object that always works :/

Comment (beckernick) [reply dask/dataframe/core.py]: Good catch. Will think about this.

Comment (beckernick) [reply dask/dataframe/core.py]: Could we leverage the fact that slices behave differently? Instead of grabbing with `a.iloc[0]` let's perhaps grab with `a.iloc[:0]` or `a.iloc[0:1]`? This appears to solve the indexing issue but results in a downstream issue still. Looking into it.

Comment (beckernick) [reply dask/dataframe/core.py]: Also, as a note, it looks like the original `_take_last` also fails when there are empty partitions.

```python
...
/Users/nickbecker/NVIDIA/dask/dask/dataframe/core.py in _take_last(a, skipna)
   4213         last_row = a.groupby(group_dummy).last()
   4214         if isinstance(a, pd.DataFrame):  # TODO: handle explicit pandas reference
-> 4215             return pd.Series(last_row.values[0], index=a.columns)
   4216         else:
   4217             return last_row.values[0]

IndexError: index 0 is out of bounds for axis 0 with size 0
```

Comment (mrocklin) [reply dask/dataframe/core.py]: What if it has no columns?

Comment (mrocklin) [reply dask/dataframe/core.py]: If this failed before then we can probably leave this as is and raise an issue.

Comment (mrocklin) [reply dask/dataframe/core.py]: Ah, seems like you just did this in https://github.com/dask/dask/issues/4743

**Comment by mrocklin** [comment]
Merging this tomorrow if there are no further comments

File: dask/dataframe/core.py (hunk 7 above)
Comment (beckernick) [reply dask/dataframe/core.py]: Yep. Defer to you on whether you want to leave this as is since it was already failing on empty partition dataframes or try to fix it in this PR.


The above snippet does solve the problem for cumsum and cumprod, but still causes issues with cummin and cummax when it tries to compare a non-empty series with an empty series:

```python
/dask/dask/dataframe/methods.py in cummin_aggregate(x, y)
    147 def cummin_aggregate(x, y):
    148     if isinstance(x, (pd.Series, pd.DataFrame)):
--> 149         return x.where((x < y) | x.isnull(), y, axis=x.ndim - 1)
    150     else:       # scalar
    151         return x if x < y else y

/conda/envs/cudf/lib/python3.7/site-packages/pandas/core/ops.py in wrapper(self, other, axis)
   1674
   1675         elif isinstance(other, ABCSeries) and not self._indexed_same(other):
-> 1676             raise ValueError("Can only compare identically-labeled "
   1677                              "Series objects")
   1678

ValueError: Can only compare identically-labeled Series objects
ipdb> self
x    0
dtype: int64
ipdb> other
Series([], dtype: float64)
ipdb>
```

Comment (mrocklin) [reply dask/dataframe/core.py]: I'd be happy if you wanted to include it.  You might also include a test with the currently passing cum-reductions (and, if you wanted to get fancy, xfailed parameters for the failing ones).

```python
@pytest.mark.parametrize('func', [
    M.cumsum,
    M.cumprod,
    pytest.param(M.cummin, marks=[pytest.mark.xfail(reason="...")]),
    ...
])
def test_cum_reduction_empty_partitions(func):
    df = ...
    ...
```

(I could be totally wrong about the `pytest.param` syntax by the way.  I would check existing tests.

Comment (beckernick) [reply dask/dataframe/core.py]: I'll include the snippet and add tests for the empty partition case so that we still can at least pass cumsum/cumprod with empty partitions, which is an improvement on the current situation. I'll match the xfail style for the cummin/cummax on empty partitions to the style in the other tests.

**Comment by mrocklin** [comment]
This is in.  Thanks @beckernick !
